promql/parse_test.go

```diff
@@ -122,39 +124,77 @@ var testExpr = []struct {
 			},
 		},
 	}, {
-		input: "", fail: true,
+		input:  "",
+		fail:   true,
+		errMsg: "no expression found in input",
+	}, {
+		input:  "# just a comment\n\n",
+		fail:   true,
+		errMsg: "no expression found in input",
 	}, {
-		input: "# just a comment\n\n", fail: true,
+		input:  "1+",
+		fail:   true,
+		errMsg: "could not find right-hand side of binary expression",
 	}, {
-		input: "1+", fail: true,
+		input:  ".",
+		fail:   true,
+		errMsg: "unrecognized character in statement: '.'",
```

Comment:
The lexer has no notion of expression vs statement. It's a question of how much complexity/state we want to introduce for more specific error messages.

At some points it even makes sense to let the lex stuff we know to be wrong to have more state for errors in the parser - it's a fine line.